iroh-relay/src/server/client_conn.rs

```diff
@@ -317,10 +327,166 @@ impl Actor {
+        item: anyhow::Result<Frame>,
+    },
+    Ready,
+}
+
+impl RateLimitedRelayedStream {
+    fn new(inner: RelayedStream, limiter: governor::DefaultDirectRateLimiter) -> Self {
+        Self {
+            inner,
+            limiter: Arc::new(limiter),
+            state: State::Ready,
+            limited_once: false,
+        }
+    }
+}
+
+impl RateLimitedRelayedStream {
+    /// Records metrics about being rate-limited.
+    fn record_rate_limited(&mut self) {
+        // TODO: add a label for the frame type.
+        inc!(Metrics, frames_rx_ratelimited_total);
+        if !self.limited_once {
+            inc!(Metrics, conns_rx_ratelimited_total);
+            self.limited_once = true;
+        }
+    }
+}
+
+impl Stream for RateLimitedRelayedStream {
+    type Item = anyhow::Result<Frame>;
+
+    #[instrument(name = "rate_limited_relayed_stream", skip_all)]
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        loop {
+            match &mut self.state {
+                State::Ready => {
+                    // Poll inner for a new item.
+                    match Pin::new(&mut self.inner).poll_next(cx) {
+                        Poll::Ready(Some(item)) => {
+                            match &item {
+                                Ok(frame) => {
+                                    // How many bytes does this frame consume?
+                                    let Ok(frame_len) =
+                                        TryInto::<u32>::try_into(frame.len_with_header())
+                                            .and_then(TryInto::<NonZeroU32>::try_into)
+                                    else {
+                                        error!("frame len not NonZeroU32, is MAX_FRAME_SIZE too large?");
+                                        // Let this frame through so to not completely break.
+                                        return Poll::Ready(Some(item));
+                                    };
+
+                                    match self.limiter.check_n(frame_len) {
+                                        Ok(Ok(_)) => return Poll::Ready(Some(item)),
+                                        Ok(Err(_)) => {
+                                            // Item is rate-limited.
+                                            self.record_rate_limited();
+                                            let limiter = self.limiter.clone();
+                                            let delay = Box::pin(async move {
+                                                limiter.until_n_ready(frame_len).await.ok();
+                                            });
+                                            self.state = State::Blocked { delay, item };
+                                            continue;
+                                        }
+                                        Err(_insufficient_capacity) => {
```

Comment:
how could this happen?

Reply:
If someone configures a max_burst_bytes smaller than the MAX_FRAME_SIZE.  So this is effectively a mis-configuration, hence I opted to keep the relay functional and leave the frame through but log an error.

I've improved the error message to explain this a bit more.